Pathio test failure on FreeBSD

```
============================= test session starts ==============================
platform freebsd12 -- Python 3.8.3, pytest-4.5.0, py-1.8.1, pluggy-0.13.1
rootdir: /ssd/portstrees/batchports-mem/ftp/py-aioftp/work-py38/aioftp-0.16.1
plugins: asyncio-0.10.0, trio-0.6.0, cov-2.9.0, hypothesis-5.19.0
collected 767 items

tests/test_abort.py ..........                                           [  1%]
tests/test_client_side_socks.py ........                                 [  2%]
tests/test_connection.py ..........................................      [  7%]
tests/test_corner_cases.py ..............                                [  9%]
tests/test_current_directory.py ........                                 [ 10%]
tests/test_directory_actions.py ..................                       [ 13%]
tests/test_extra.py .......                                              [ 13%]
tests/test_file.py ........................................              [ 19%]
tests/test_list_fallback.py .........                                    [ 20%]
tests/test_login.py ................                                     [ 22%]
tests/test_maximum_connections.py ..........                             [ 23%]
tests/test_passive.py ....................                               [ 26%]
tests/test_pathio.py ................................................... [ 32%]
...................FF........................                            [ 38%]
tests/test_permissions.py ....                                           [ 39%]
tests/test_restart.py ......................                             [ 42%]
tests/test_simple_functions.py ......................................... [ 47%]
........................................................................ [ 56%]
........................................................................ [ 66%]
........................................................................ [ 75%]
................                                                         [ 77%]
tests/test_throttle.py ................................................. [ 84%]
........................................................................ [ 93%]
................................................                         [ 99%]
tests/test_user.py .                                                     [100%]

=================================== FAILURES ===================================
__________________________ test_unlink_on_dir[PathIO] __________________________

args = (<aioftp.pathio.PathIO object at 0x8040f63a0>, PosixPath('/tmp/tmplcfgnik8'))
kwargs = {}

    @functools.wraps(coro)
    async def wrapper(*args, **kwargs):
        try:
>           return await coro(*args, **kwargs)

aioftp/pathio.py:80: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <aioftp.pathio.PathIO object at 0x8040f63a0>
path = PosixPath('/tmp/tmplcfgnik8')

    @universal_exception
    async def unlink(self, path):
>       return path.unlink()

aioftp/pathio.py:402: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = PosixPath('/tmp/tmplcfgnik8'), missing_ok = False

    def unlink(self, missing_ok=False):
        """
        Remove this file or link.
        If the path is a directory, use rmdir() instead.
        """
        if self._closed:
            self._raise_closed()
        try:
>           self._accessor.unlink(self)
E           PermissionError: [Errno 1] Operation not permitted: '/tmp/tmplcfgnik8'

/usr/local/lib/python3.8/pathlib.py:1321: PermissionError

During handling of the above exception, another exception occurred:

exc = (<class 'IsADirectoryError'>,), type = <class 'PermissionError'>
instance = PermissionError(1, 'Operation not permitted')

    @contextlib.contextmanager
    def universal_exception_reason(*exc):
        try:
>           yield

tests/test_pathio.py:11: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

path_io = <aioftp.pathio.PathIO object at 0x8040f63a0>
temp_dir = PosixPath('/tmp/tmplcfgnik8')

    @pytest.mark.asyncio
    async def test_unlink_on_dir(path_io, temp_dir):
        with universal_exception_reason(IsADirectoryError):
>           await path_io.unlink(temp_dir)

tests/test_pathio.py:256: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<aioftp.pathio.PathIO object at 0x8040f63a0>, PosixPath('/tmp/tmplcfgnik8'))
kwargs = {}

    @functools.wraps(coro)
    async def wrapper(*args, **kwargs):
        try:
            return await coro(*args, **kwargs)
        except (asyncio.CancelledError, NotImplementedError,
                StopAsyncIteration):
            raise
        except Exception:
>           raise errors.PathIOError(reason=sys.exc_info())
E           aioftp.errors.PathIOError

aioftp/pathio.py:85: PathIOError

During handling of the above exception, another exception occurred:

path_io = <aioftp.pathio.PathIO object at 0x8040f63a0>
temp_dir = PosixPath('/tmp/tmplcfgnik8')

    @pytest.mark.asyncio
    async def test_unlink_on_dir(path_io, temp_dir):
        with universal_exception_reason(IsADirectoryError):
>           await path_io.unlink(temp_dir)

tests/test_pathio.py:256: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.8/contextlib.py:131: in __exit__
    self.gen.throw(type, value, traceback)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

exc = (<class 'IsADirectoryError'>,), type = <class 'PermissionError'>
instance = PermissionError(1, 'Operation not permitted')

    @contextlib.contextmanager
    def universal_exception_reason(*exc):
        try:
            yield
        except aioftp.PathIOError as e:
            type, instance, traceback = e.reason
            m = f"Expect one of {exc}, got {instance}"
>           assert isinstance(instance, exc), m
E           AssertionError: Expect one of (<class 'IsADirectoryError'>,), got [Errno 1] Operation not permitted: '/tmp/tmplcfgnik8'
E           assert False
E            +  where False = isinstance(PermissionError(1, 'Operation not permitted'), (<class 'IsADirectoryError'>,))

tests/test_pathio.py:15: AssertionError
_______________________ test_unlink_on_dir[AsyncPathIO] ________________________

args = (<aioftp.pathio.AsyncPathIO object at 0x80440d8b0>, PosixPath('/tmp/tmpbazjjqrd'))
kwargs = {}

    @functools.wraps(coro)
    async def wrapper(*args, **kwargs):
        try:
>           return await coro(*args, **kwargs)

aioftp/pathio.py:80: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

fut = <coroutine object AsyncPathIO.unlink at 0x803f3fac0>, timeout = None

    async def wait_for(fut, timeout, *, loop=None):
        """Wait for the single Future or coroutine to complete, with timeout.
    
        Coroutine will be wrapped in Task.
    
        Returns result of the Future or coroutine.  When a timeout occurs,
        it cancels the task and raises TimeoutError.  To avoid the task
        cancellation, wrap it in shield().
    
        If the wait is cancelled, the task is also cancelled.
    
        This function is a coroutine.
        """
        if loop is None:
            loop = events.get_running_loop()
        else:
            warnings.warn("The loop argument is deprecated since Python 3.8, "
                          "and scheduled for removal in Python 3.10.",
                          DeprecationWarning, stacklevel=2)
    
        if timeout is None:
>           return await fut

/usr/local/lib/python3.8/asyncio/tasks.py:455: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <aioftp.pathio.AsyncPathIO object at 0x80440d8b0>
args = (PosixPath('/tmp/tmpbazjjqrd'),), kwargs = {}

    @functools.wraps(f)
    async def wrapper(self, *args, **kwargs):
>       return await asyncio.get_event_loop().run_in_executor(
            self.executor,
            functools.partial(f, self, *args, **kwargs),
        )

aioftp/pathio.py:456: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = None

    def run(self):
        if not self.future.set_running_or_notify_cancel():
            return
    
        try:
>           result = self.fn(*self.args, **self.kwargs)

/usr/local/lib/python3.8/concurrent/futures/thread.py:57: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <aioftp.pathio.AsyncPathIO object at 0x80440d8b0>
path = PosixPath('/tmp/tmpbazjjqrd')

    @universal_exception
    @with_timeout
    @_blocking_io
    def unlink(self, path):
>       return path.unlink()

aioftp/pathio.py:511: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = PosixPath('/tmp/tmpbazjjqrd'), missing_ok = False

    def unlink(self, missing_ok=False):
        """
        Remove this file or link.
        If the path is a directory, use rmdir() instead.
        """
        if self._closed:
            self._raise_closed()
        try:
>           self._accessor.unlink(self)
E           PermissionError: [Errno 1] Operation not permitted: '/tmp/tmpbazjjqrd'

/usr/local/lib/python3.8/pathlib.py:1321: PermissionError

During handling of the above exception, another exception occurred:

exc = (<class 'IsADirectoryError'>,), type = <class 'PermissionError'>
instance = PermissionError(1, 'Operation not permitted')

    @contextlib.contextmanager
    def universal_exception_reason(*exc):
        try:
>           yield

tests/test_pathio.py:11: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

path_io = <aioftp.pathio.AsyncPathIO object at 0x80440d8b0>
temp_dir = PosixPath('/tmp/tmpbazjjqrd')

    @pytest.mark.asyncio
    async def test_unlink_on_dir(path_io, temp_dir):
        with universal_exception_reason(IsADirectoryError):
>           await path_io.unlink(temp_dir)

tests/test_pathio.py:256: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<aioftp.pathio.AsyncPathIO object at 0x80440d8b0>, PosixPath('/tmp/tmpbazjjqrd'))
kwargs = {}

    @functools.wraps(coro)
    async def wrapper(*args, **kwargs):
        try:
            return await coro(*args, **kwargs)
        except (asyncio.CancelledError, NotImplementedError,
                StopAsyncIteration):
            raise
        except Exception:
>           raise errors.PathIOError(reason=sys.exc_info())
E           aioftp.errors.PathIOError

aioftp/pathio.py:85: PathIOError

During handling of the above exception, another exception occurred:

path_io = <aioftp.pathio.AsyncPathIO object at 0x80440d8b0>
temp_dir = PosixPath('/tmp/tmpbazjjqrd')

    @pytest.mark.asyncio
    async def test_unlink_on_dir(path_io, temp_dir):
        with universal_exception_reason(IsADirectoryError):
>           await path_io.unlink(temp_dir)

tests/test_pathio.py:256: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.8/contextlib.py:131: in __exit__
    self.gen.throw(type, value, traceback)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

exc = (<class 'IsADirectoryError'>,), type = <class 'PermissionError'>
instance = PermissionError(1, 'Operation not permitted')

    @contextlib.contextmanager
    def universal_exception_reason(*exc):
        try:
            yield
        except aioftp.PathIOError as e:
            type, instance, traceback = e.reason
            m = f"Expect one of {exc}, got {instance}"
>           assert isinstance(instance, exc), m
E           AssertionError: Expect one of (<class 'IsADirectoryError'>,), got [Errno 1] Operation not permitted: '/tmp/tmpbazjjqrd'
E           assert False
E            +  where False = isinstance(PermissionError(1, 'Operation not permitted'), (<class 'IsADirectoryError'>,))

tests/test_pathio.py:15: AssertionError
==================== 2 failed, 765 passed in 41.11 seconds =====================
*** Error code 1
```

--- a/history.rst
+++ b/history.rst
@@ -2,6 +2,7 @@ x.x.x (xx-xx-xxxx)
 ------------------
 
 - tests: fix test_unlink_on_dir on POSIX compatible systems (#118)
+Thanks to `AMDmi3 <https://github.com/AMDmi3>`_
 
 0.16.1 (09-07-2020)
 ------------------

--- a/history.rst
+++ b/history.rst
@@ -1,6 +1,8 @@
 x.x.x (xx-xx-xxxx)
 ------------------
 
+- tests: fix test_unlink_on_dir on POSIX compatible systems (#118)
+
 0.16.1 (09-07-2020)
 ------------------
 

--- a/tests/test_pathio.py
+++ b/tests/test_pathio.py
@@ -252,7 +252,7 @@ async def test_unlink_not_exist(path_io, temp_dir):
 
 @pytest.mark.asyncio
 async def test_unlink_on_dir(path_io, temp_dir):
-    with universal_exception_reason(IsADirectoryError):
+    with universal_exception_reason(IsADirectoryError, PermissionError):
         await path_io.unlink(temp_dir)
 
 

